sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/MergeScalarSubqueriesSuite.scala

```diff
@@ -0,0 +1,355 @@
+      .select(
+        CreateNamedStruct(Seq(
+          Literal("a_plus1"), 'a_plus1,
+          Literal("a_plus2"), 'a_plus2,
+          Literal("b"), 'b
+        )).as("mergedValue"))
+    val analyzedMergedSubquery = mergedSubquery.analyze
+    val correctAnswer = WithCTE(
+      testRelation
+        .select(
+          extractorExpression(0, analyzedMergedSubquery.output, 0),
+          extractorExpression(0, analyzedMergedSubquery.output, 1),
+          extractorExpression(0, analyzedMergedSubquery.output, 2),
+          extractorExpression(0, analyzedMergedSubquery.output, 0),
+          extractorExpression(0, analyzedMergedSubquery.output, 1),
+          extractorExpression(0, analyzedMergedSubquery.output, 2)),
+      Seq(CTERelationDef(analyzedMergedSubquery, 0)))
+
+    comparePlans(Optimize.execute(originalQuery.analyze), correctAnswer.analyze)
+  }
+
+  test("Merging subqueries with aggregates") {
+    val subquery1 = ScalarSubquery(testRelation.groupBy('b)(max('a).as("max_a")))
+    val subquery2 = ScalarSubquery(testRelation.groupBy('b)(sum('a).as("sum_a")))
+    val subquery3 = ScalarSubquery(testRelation.groupBy('b)('b))
+    val subquery4 = ScalarSubquery(testRelation.groupBy('b)(max('a).as("max_a_2")))
+    val subquery5 = ScalarSubquery(testRelation.groupBy('b)(sum('a).as("sum_a_2")))
+    val subquery6 = ScalarSubquery(testRelation.groupBy('b)('b.as("b_2")))
+    val originalQuery = testRelation
+      .select(
+        subquery1,
+        subquery2,
+        subquery3,
+        subquery4,
+        subquery5,
+        subquery6)
+
+    val mergedSubquery = testRelation
+      .groupBy('b)(
+        max('a).as("max_a"),
+        sum('a).as("sum_a"),
+        'b)
+      .select(CreateNamedStruct(Seq(
+        Literal("max_a"), 'max_a,
+        Literal("sum_a"), 'sum_a,
+        Literal("b"), 'b
+      )).as("mergedValue"))
+    val analyzedMergedSubquery = mergedSubquery.analyze
+    val correctAnswer = WithCTE(
+      testRelation
+        .select(
+          extractorExpression(0, analyzedMergedSubquery.output, 0),
+          extractorExpression(0, analyzedMergedSubquery.output, 1),
+          extractorExpression(0, analyzedMergedSubquery.output, 2),
+          extractorExpression(0, analyzedMergedSubquery.output, 0),
+          extractorExpression(0, analyzedMergedSubquery.output, 1),
+          extractorExpression(0, analyzedMergedSubquery.output, 2)),
+      Seq(CTERelationDef(analyzedMergedSubquery, 0)))
+
+    comparePlans(Optimize.execute(originalQuery.analyze), correctAnswer.analyze)
+  }
+
+  test("Merging subqueries with filters") {
+    val subquery1 = ScalarSubquery(testRelation.where('a > 1).select('a))
+    // Despite having an extra Project node, `subquery2` is mergeable with `subquery1`
+    val subquery2 = ScalarSubquery(testRelation.where('a > 1).select('b.as("b_1")).select('b_1))
+    // Despite lacking a Project node, `subquery3` is mergeable with the result of merging
+    // `subquery1` and `subquery2`
```

Comment:
Add a test case to let subquery3 and subquery4 have different filter conditions?

Reply:
Added a new test: https://github.com/apache/spark/pull/32298/commits/1ff64e461ea5c710660121727dabf77b1318a587